crates/revm/src/evm_impl.rs

```diff
@@ -207,11 +272,61 @@ impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> Transact<DB::Error>
         if crate::USE_GAS {
             match exit_reason {
                 return_ok!() => {
-                    gas.erase_cost(ret_gas.remaining());
-                    gas.record_refund(ret_gas.refunded());
+                    #[cfg(not(feature = "optimism"))]
+                    {
+                        gas.erase_cost(ret_gas.remaining());
+                        gas.record_refund(ret_gas.refunded());
+                    }
+
+                    // On Optimism, deposit transactions report gas usage uniquely to other
+                    // transactions due to them being pre-paid on L1.
+                    //
+                    // Hardfork Behavior:
+                    // - Bedrock (success path):
+                    //   - Deposit transactions (non-system) report their gas limit as the usage.
+                    //     No refunds.
+                    //   - Deposit transactions (system) report 0 gas used. No refunds.
+                    //   - Regular transactions report gas usage as normal.
+                    // - Regolith (success path):
+                    //   - Deposit transactions (all) report their gas used as normal. Refunds
+                    //     enabled.
+                    //   - Regular transactions report their gas used as normal.
+                    #[cfg(feature = "optimism")]
+                    if self.data.env.cfg.optimism
+                        && (!is_deposit || GSPEC::enabled(SpecId::REGOLITH))
+                    {
+                        // For regular transactions prior to Regolith and all transactions after
+                        // Regolith, gas is reported as normal.
+                        gas.erase_cost(ret_gas.remaining());
+                        gas.record_refund(ret_gas.refunded());
+                    } else if is_deposit && tx_system.unwrap_or(false) {
+                        // System transactions were a special type of deposit transaction in
+                        // the Bedrock hardfork that did not incur any gas costs.
+                        gas.erase_cost(tx_gas_limit);
+                    }
```

Comment:
This is the case when the transaction is a deposit transaction pre-regolith. In this case (as clabby noted in the comments), the deposit transaction reports the gas limit as its usage since it's pre-paid for on L1. As such, we don't need to erase any gas costs or perform any refunds :)